### PR TITLE
fix: upcast reranker logits to float32 before softmax (HPS)

### PR DESCRIPTION
**Pull request by KisuYang**
## What
The Qwen3 and monoT5-family cross-encoder rerankers now upcast their logits to **float32** immediately before `log_softmax`.

## Why
These rerankers run the model in **bfloat16/float16** and applied the softmax directly to the low-precision logits. The reduced mantissa coarsely buckets the (0, 1) probability range, so distinct query–document relevance scores collapse into **spurious ties**. Reranking metrics then depend on arbitrary tie-breaking, which inflates their variance and bias — e.g. for Qwen3-Reranker-0.6B under bf16 the paper reports up to ~38%p MRR@10 range and +9.08%p bias, both largely removed by this one-line **High-Precision Scoring (HPS)** upcast (Yang et al., 2026, https://aclanthology.org/2026.acl-short.33).

The upcast is value-preserving and adds negligible cost (only a small logits tensor is converted). `kalm_reranker.py` already does this; this PR brings the other softmax rerankers in line.

## Changes
- `mteb/models/model_implementations/qwen3_reranker.py`: `.float()` before `log_softmax` in `compute_logits`.
- `mteb/models/model_implementations/rerankers_monot5_based.py`: `.float()` before each `log_softmax` (MonoT5Reranker and LlamaReranker paths).

## Checks
- `ruff check` / `ruff format --check`: pass on changed files
- Modules compile; scoring math verified: a bf16 two-logit softmax over 200 candidates yields ~145 unique scores, vs ~196 with HPS (fp32 reference: 200).

## Notes
Disclosure: I'm the first author of the referenced paper. Companion PR handles the bi-encoder similarity primitives. Related issue: #4932.

### CODE REVIEW

File: mteb/models/model_implementations/qwen3_reranker.py [hunk 1]
```diff
@@ -101,7 +101,13 @@ def compute_logits(self, inputs: dict) -> list[float]:
         true_vector = batch_scores[:, self.token_true_id]
         false_vector = batch_scores[:, self.token_false_id]
         batch_scores = torch.stack([false_vector, true_vector], dim=1)
-        batch_scores = torch.nn.functional.log_softmax(batch_scores, dim=1)
+        # High-Precision Scoring (HPS): upcast the logits to float32 before the
+        # softmax. The model runs in bfloat16, whose 7-bit mantissa coarsely
+        # buckets the (0, 1) probability range and collapses distinct relevance
+        # scores into spurious ties. The upcast is value-preserving and adds
+        # negligible cost. See "Reliable Evaluation Protocol for Low-Precision
+        # Retrieval" (Yang et al., 2026), https://aclanthology.org/2026.acl-short.33.
```
Comment (Samoed) [inline mteb/models/model_implementations/qwen3_reranker.py]: Can you shorten comments a bit
```suggestion
        # upcast the logits to float32 before the softmax
```

**Comment by Samoed** [comment]
Can you shorten comments in other places?

**Comment by KisuYang** [comment]
> Can you shorten comments in other places?

Sure, I will shorten other comments as well.

**Comment by KisuYang** [comment]
Hi [Samoed](https://github.com/Samoed), I've addressed your comments. Could you please take another look, and merge if it looks good to you? Thanks!

**Review by KennethEnevoldsen — APPROVED** [review]
I think this is good to merge thanks for the fix @KisuYang and nice paper outlining the issue

@Samoed anything you are concerned about here? (We could consider doing this always doing the upcast)
